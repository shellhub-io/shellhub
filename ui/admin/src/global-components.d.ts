// src/global-components.d.ts
import SnackbarComponent from "./components/Snackbar/Snackbar.vue";

declare module "@vue/runtime-core" {
  export interface GlobalComponents {
    SnackbarComponent: typeof SnackbarComponent;
    RouterLink: typeof import("vue-router")["RouterLink"];
    RouterView: typeof import("vue-router")["RouterView"];
  }
}
