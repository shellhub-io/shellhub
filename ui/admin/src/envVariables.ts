/* eslint-disable import/prefer-default-export */
export const envVariables = {
  announcementsEnable: process.env.SHELLHUB_ANNOUNCEMENTS === "true",
  tinyMceKey: process.env.TINYMCE_API_KEY,
};
