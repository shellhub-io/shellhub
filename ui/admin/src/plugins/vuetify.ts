// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";
import { VNumberInput } from "vuetify/labs/VNumberInput";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
  components: {
    VNumberInput,
  },
});
