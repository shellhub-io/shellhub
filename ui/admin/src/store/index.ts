import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
