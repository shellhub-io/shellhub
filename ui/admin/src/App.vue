<template>
  <v-app>
    <component :is="layout" :data-test="`${layout}-component`" />
  </v-app>
</template>

<script lang="ts">
import { defineComponent, onMounted, computed } from "vue";
import { useRouter } from "vue-router";
import SimpleLayout from "./layouts/SimpleLayout.vue";
import AppLayout from "./layouts/AppLayout.vue";
import { useStore } from "./store";
import { INotificationsError } from "./interfaces/INotifications";

export default defineComponent({
  name: "App",
  components: {
    appLayout: AppLayout,
    simpleLayout: SimpleLayout,
  },
  setup() {
    const store = useStore();
    const router = useRouter();

    const layout = computed(() => store.getters["layout/getLayout"]);

    const isLoggedIn = computed(() => store.getters["auth/isLoggedIn"]);

    const currentRoute = computed(() => router.currentRoute.value.path);

    onMounted(async () => {
      if (!isLoggedIn.value) {
        try {
          await store.dispatch("auth/logout");
          store.dispatch("layout/setLayout", "simpleLayout");
          router.push("/login");
        } catch {
          store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.namespaceLoad);
        }
      }

      if (isLoggedIn.value && currentRoute.value !== "/login") {
        const license = await store.dispatch("license/get");
        if (!license || license.expired) {
          store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.license);
          store.dispatch("layout/setLayout", "appLayout");
          router.push("/license");
        }
        store.dispatch("layout/setLayout", "appLayout");
      }
    });

    return {
      layout,
    };
  },
});
</script>
