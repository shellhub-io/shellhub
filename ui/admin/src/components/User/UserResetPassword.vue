<template>
  <v-tooltip bottom anchor="bottom">
    <template v-slot:activator="{ props }">
      <v-icon
        tag="a"
        dark
        v-bind="props"
        @click="dialog = true"
        tabindex="0"
        data-test="open-dialog-icon"
      >mdi-account-lock-open
      </v-icon>
    </template>
    <span>Enable Local Authentication</span>
  </v-tooltip>

  <v-dialog v-model="dialog" @click:outside="close" max-width="500">
    <v-card>
      <v-card-title class="text-h5 pb-2">Enable Local Authentication</v-card-title>
      <v-divider />
      <v-window v-model="step">
        <v-window-item value="step-1">
          <v-card-text class="pb-0" data-test="step-1-description">
            This action will enable local authentication to this user and generate a new password.
          </v-card-text>

          <v-card-text data-test="step-1-confirmation">
            Do you want to enable?
          </v-card-text>

          <v-card-actions>
            <v-btn @click="close" data-test="cancel-btn">Cancel</v-btn>
            <v-spacer />
            <v-btn color="primary" @click="proceedToSecondStep" data-test="enable-btn">Enable</v-btn>
          </v-card-actions>
        </v-window-item>

        <v-window-item value="step-2">
          <v-card-text>
            <v-alert
              class="mb-2"
              type="warning"
              variant="tonal"
              text="Users are strongly encouraged to change this password after their first successful authentication"
              data-test="password-warning"
            />
            A new password has been generated for this user. Please copy it and share it securely:
            <v-text-field
              class="mt-4"
              v-model="generatedPassword"
              readonly
              @click="copyText(generatedPassword)"
              @keypress="copyText(generatedPassword)"
              variant="outlined"
              dense
              prepend-inner-icon="mdi-key"
              append-inner-icon="mdi-clipboard-text"
              data-test="generated-password-field"
            />
          </v-card-text>

          <v-card-actions>
            <v-btn @click="close" data-test="close-btn">Close</v-btn>
          </v-card-actions>
        </v-window-item>
      </v-window>
    </v-card>
  </v-dialog>
</template>

<script setup lang="ts">
import { ref, computed } from "vue";
import { useStore } from "../../store";
import { INotificationsCopy } from "../../interfaces/INotifications";

const props = defineProps({
  userId: {
    type: String,
    default: "",
    required: true,
  },
});

const emit = defineEmits(["update"]);

const store = useStore();
const dialog = ref(false);
const step = ref("step-1");
const generatedPassword = computed(() => store.getters["users/getGeneratedPassword"]);

const close = () => {
  dialog.value = false;
  step.value = "step-1";
  emit("update");
};

const copyText = (value: string | undefined) => {
  if (value) {
    navigator.clipboard.writeText(value);
    store.dispatch("snackbar/showSnackbarCopy", INotificationsCopy.tenantId);
  }
};

const proceedToSecondStep = async () => {
  try {
    await store.dispatch("users/resetUserPassword", props.userId);
    step.value = "step-2";
  } catch (error) {
    store.dispatch(
      "snackbar/showSnackbarErrorAction",
      "Failed to reset user password. Please try again.",
    );
  }
};

defineExpose({ dialog, step });
</script>
