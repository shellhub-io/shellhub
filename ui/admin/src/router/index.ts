import { createRouter, createWebHistory } from "vue-router";
import Login from "@admin/views/Login.vue";
import Dashboard from "@admin/views/Dashboard.vue";
import Users from "@admin/views/Users.vue";
import SettingsLicense from "@admin/components/Settings/SettingsLicense.vue";
import SettingsAuthentication from "@admin/components/Settings/SettingsAuthentication.vue";
import Namespaces from "@admin/views/Namespaces.vue";
import Settings from "@admin/views/Settings.vue";

import { store } from "../store";

const routes = [
  {
    path: "/login",
    name: "login",
    component: Login,
  },
  {
    path: "/",
    name: "dashboard",
    component: Dashboard,
  },
  {
    path: "/users",
    name: "users",
    component: Users,
  },
  {
    path: "/user/:id",
    name: "userDetails",
    component: () => import("@admin/views/UserDetails.vue"),
  },
  {
    path: "/devices",
    name: "devices",
    component: () => import("@admin/views/Device.vue"),
    redirect: {
      name: "listDevices",
    },
    children: [
      {
        path: "",
        name: "listDevices",
        component: () => import("@admin/views/Device.vue"),
      },
    ],
  },
  {
    path: "/device/:id",
    name: "deviceDetails",
    component: () => import("@admin/views/DeviceDetails.vue"),
  },
  {
    path: "/sessions",
    name: "sessions",
    component: () => import("@admin/views/Sessions.vue"),
  },
  {
    path: "/session/:id",
    name: "sessionDetails",
    component: () => import("@admin/views/SessionDetails.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    component: Settings,
    redirect: { name: "SettingProfile" },
    children: [
      {
        path: "authentication",
        name: "SettingAuthentication",
        component: SettingsAuthentication,
      },
      {
        path: "license",
        name: "SettingLicense",
        component: SettingsLicense,
      },
    ],
  },
  {
    path: "/firewall-rules",
    name: "firewall-rules",
    component: () => import("@admin/views/FirewallRules.vue"),
  },
  {
    path: "/firewall-rules/:id",
    name: "firewallRulesDetails",
    component: () => import("@admin/views/FirewallRulesDetails.vue"),
  },
  {
    path: "/namespaces",
    name: "namespaces",
    component: Namespaces,
  },
  {
    path: "/namespace/:id",
    name: "namespaceDetails",
    component: () => import("@admin/views/NamespaceDetails.vue"),
  },
  {
    path: "/announcements",
    name: "announcements",
    component: () => import("@admin/views/Announcements.vue"),
  },
  {
    path: "/announcement/:uuid",
    name: "announcementDetails",
    component: () => import("@admin/views/AnnouncementDetails.vue"),
  },
  {
    path: "/new-announcement",
    name: "new-announcement",
    component: () => import("@admin/views/NewAnnouncement.vue"),
  },
];

const router = createRouter({
  history: createWebHistory("/admin/"),
  routes,
});

router.beforeEach((to, from, next) => {
  if (to.path !== "/login") {
    if (store.getters["auth/isLoggedIn"]) {
      return next();
    }
    return next(`/login?redirect=${to.path}`);
  }
  if (store.getters["auth/isLoggedIn"]) {
    if (to.path === "/login" && to.query.token) {
      return next();
    }
    return next("/");
  }

  return next();
});

export default router;
