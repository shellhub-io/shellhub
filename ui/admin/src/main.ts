import { createApp } from "vue";
import vuetify from "@ui/plugins/vuetify";
import App from "./App.vue";
import { loadFonts } from "./plugins/webfontloader";
import { key, store } from "./store";
import router from "./router";
import SnackbarComponent from "./components/Snackbar/Snackbar.vue";

const app = createApp(App);

loadFonts();

app.use(vuetify);
app.use(router);
app.use(store, key);
app.component("SnackbarComponent", SnackbarComponent);
app.mount("#app");
