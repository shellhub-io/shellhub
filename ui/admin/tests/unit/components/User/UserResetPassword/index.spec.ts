import { beforeEach, describe, it, expect, vi } from "vitest";
import { flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { createVuetify } from "vuetify";
import UserResetPassword from "../../../../../src/components/User/UserResetPassword.vue";
import { store, key } from "../../../../../src/store";

type UserResetPasswordWrapper = VueWrapper<InstanceType<typeof UserResetPassword>>;

describe("User Reset Password", () => {
  let wrapper: UserResetPasswordWrapper;

  const vuetify = createVuetify();

  const mockProps = {
    userId: "user123",
  };

  beforeEach(() => {
    wrapper = mount(UserResetPassword, {
      global: {
        plugins: [[store, key], vuetify],
      },
      props: mockProps,
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("opens the dialog when the icon is clicked", async () => {
    await wrapper.findComponent("[data-test='open-dialog-icon']").trigger("click");
    expect(wrapper.vm.dialog).toBe(true);
  });

  it("closes the dialog when 'Close' is clicked", async () => {
    await wrapper.findComponent("[data-test='open-dialog-icon']").trigger("click");
    await flushPromises();
    await wrapper.findComponent("[data-test='cancel-btn']").trigger("click");
    expect(wrapper.vm.dialog).toBe(false);
    expect(wrapper.vm.step).toBe("step-1");
  });

  it("proceeds to the second step when 'Enable' is clicked", async () => {
    await wrapper.findComponent("[data-test='open-dialog-icon']").trigger("click");

    const resetUserPasswordSpy = vi.spyOn(store, "dispatch").mockResolvedValueOnce({});

    await wrapper.findComponent("[data-test='enable-btn']").trigger("click");
    expect(resetUserPasswordSpy).toHaveBeenCalledWith("users/resetUserPassword", mockProps.userId);
    expect(wrapper.vm.step).toBe("step-2");
  });

  it("shows an error message if 'Enable' action fails", async () => {
    await wrapper.findComponent("[data-test='open-dialog-icon']").trigger("click");

    const resetUserPasswordSpy = vi.spyOn(store, "dispatch").mockRejectedValueOnce(new Error("Error"));

    await wrapper.findComponent("[data-test='enable-btn']").trigger("click");
    expect(resetUserPasswordSpy).toHaveBeenCalledWith("users/resetUserPassword", mockProps.userId);
    expect(wrapper.vm.step).toBe("step-1");
  });

  it("resets dialog state on 'Close'", async () => {
    await wrapper.findComponent("[data-test='open-dialog-icon']").trigger("click");

    wrapper.vm.step = "step-2";

    await wrapper.findComponent("[data-test='cancel-btn']").trigger("click");
    expect(wrapper.vm.dialog).toBe(false);
    expect(wrapper.vm.step).toBe("step-1");
  });
});
