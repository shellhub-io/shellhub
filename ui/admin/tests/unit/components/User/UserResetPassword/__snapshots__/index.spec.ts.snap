// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Reset Password > renders correctly 1`] = `
"<a class="mdi-account-lock-open mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable" role="button" aria-hidden="false" tabindex="0" dark="" aria-describedby="v-tooltip-0" data-test="open-dialog-icon"></a>
<!--teleport start-->
<!--teleport end-->"
`;
